### PR TITLE
path/dynamic: revise priority queue implementation

### PR DESCRIPTION
**Pull request by kortschak**
Updates #100.

### CODE REVIEW

**Comment by kortschak** [comment]
I wrote essentially the same thing for A* and it passes for all tests except the 1000x1000 grid, which it fails by consuming far too much memory. I don't understand why this is the case, but I'll have another look tomorrow.

A quick benchmark on the D* Lite change gives moderate improvement:

```
name              old mean              new mean              delta
DStarLite         5.07ms × (1.00,1.01)  4.73ms × (1.00,1.00)    -6.67% (p=0.000 n=9+9)
DStarLiteDynamic  10.3ms × (0.99,1.00)  10.0ms × (0.99,1.02)  -2.84% (p=0.000 n=10+10)
```

Both benchmarks are on the grid in dynamicDStarLiteTests[2] either with knowledge limitation and step iteration (dynamic) or without knowledge limitation. Allocations also improve very slightly:

```
benchmark                     old allocs     new allocs     delta
BenchmarkDStarLite            9784           9771           -0.13%
BenchmarkDStarLiteDynamic     18537          18526          -0.06%

benchmark                     old bytes     new bytes     delta
BenchmarkDStarLite            526411        527302        +0.17%
BenchmarkDStarLiteDynamic     907056        906158        -0.10%
```

**Comment by kortschak** [comment]
OK. Had a quick look and I don't think this revision is possible for A* with our current implementation (which I don't think we should change); we don't have any control over the graph that we a examining and we call g.From(u.node) to get candidate v nodes. These v nodes are graph.Node at best, so we need to have some mechanism to get a priority queue index from this node - since it doesn't have an index field, we need to have an external mapping. This is our indexOf map.

So unless I've missed something here, this is ready for review.

@vladimir-ch Please take a look.

File: path/dynamic/dstarlite.go [hunk 1]
```diff
@@ -407,54 +406,63 @@ func (k key) less(other key) bool {
...
 	rhs float64
 	g   float64
 }
 
-// dStarLiteQueue is a D* Lite priority queue.
-type dStarLiteQueue struct {
-	indexOf map[int]int
-	nodes   []*dStarLiteNode
+// newDStarLiteNode returns a dStarLite node that is in a legal state
+// for existence outsite the DStarLite priority queue.
```
Comment (vladimir-ch) [inline path/dynamic/dstarlite.go]: outside

**Comment by vladimir-ch** [comment]
LGTM

**Comment by kortschak** [comment]
After I've fixed the typo are you happy this fixes 100?

**Comment by vladimir-ch** [comment]
The nodes will now all consume one int even though only a few of them are likely to ever get into the queue. But your benchmarks show (albeit minimum) improvement and the queue code is slightly cleaner, so if you are happy with this, then I am too.

**Comment by kortschak** [comment]
I'm happy with the D* Lite change, but I'm wondering if you are happy for me to close 100 without changing A* (because that's impossible with the current A* implementation - we would have to cope the input graph to do the same thing).

**Comment by vladimir-ch** [comment]
Sure, you can close #100 (I only suggested to _consider_ doing this change).

**Comment by kortschak** [comment]
Consider it considered.

**Comment by kortschak** [comment]
Closes #100.

**Pull request merged**